### ISSUE TITLE
Email address does not get updated when an existing customer is found

### ISSUE THREAD

**Issue by jacksleight** [damcclean/simple-commerce#564]
### Bug Description

This is a bit of a weird one, but I just ran into this issue. First a bit of context:

I didn't realise that Laravel's email validator allows spaces (https://github.com/laravel/framework/issues/28233). While this is probably technically correct, it's not particularly practical. Stripe however does not allow spaces, and will thrown an error if you try to submit an email address with one in.

Here's the issue:

I have an SC site where a customer mistakenly entered their email address with a space before the @ symbol, then received an error when they got to the payment gateway screen. However, once we pointed out the error and asked them to try again without the space, they continued to get the error.

This is because SC converts both `test@example.com` and `test @example.com` to the same `test-at-examplecom.md` file name, and on the subsequent attempts reuses the email value from the file rather than the submitted value.

To fix this I've manually updated the relevant customer file and implemented a `no_spaces` validator that I've added to my custom `CheckoutInformationRequest`.

```php
Validator::extend('no_spaces', function($attr, $value){
    return preg_match('/^\S*$/u', $value);
});
```

It might be worth implementing something similar in SC, especially when people are using Stripe (I haven't tested how PayPal handles this).

It would also be good if SC updated the customer data if the email has changed, even if the file name is still the same.

---

Something else that occurred to me while investigating this is that it shouldn't really be possible for multiple distinct email addresses to map to the same file. While the space thing is very much an edge case, there are other possible collisions. For example `test@example.co.uk` and `test@exampleco.uk` could in theory be distinct addresses, but would resolve to the same `test-at-examplecouk.md` file name in SC.

### Steps to reproduce

1. Submit an order with a space in the email address and try to pay with Stripe
2. Get an invalid email error from the Stripe API
3. Retry the order with the space removed
4. Get the same invalid email error from the Stripe API

### Environment

Statamic 3.2.32 Pro
Laravel 8.81.0
PHP 8.0.15
doublethreedigital/simple-commerce 2.3.64

**Comment by duncanmcclean** [comment]
Good catch - certainly an edge case!

I imagine I'll end up adding a validation rule, similar to what you've suggested, whenever an email is provided.

Regarding an email possibly mapping to the wrong file - while it's probably quite rare I agree it's something that should be fixed. I don't suppose you have any idea on how to workaround this?

**Comment by jacksleight** [comment]
I think the most straightforward thing to do would be copy how Statamic handles user files, use the entire email address, dots, @s and all. Statamic must have *some* sanitisation for those file names, but allowing any symbol permitted by the filesystem would greatly reduce the chance of a collision.

**Comment by jacksleight** [comment]
Actually looks like Statamic just uses the whole email address untouched: https://github.com/statamic/cms/blob/3943e03beae18cf9446e9d7573c8ba144f6b2993/src/Auth/File/User.php#L115

You can even add a forward slash:

![Screenshot 2022-02-15 at 12 48 55](https://user-images.githubusercontent.com/126740/154065255-7890f8c6-1d78-47d6-8979-fc2786cf1458.png)
![Screenshot 2022-02-15 at 12 46 48](https://user-images.githubusercontent.com/126740/154065110-5d97a510-fe9d-4d73-aa97-a8c1d398e1e1.png)

😬

**Comment by duncanmcclean** [comment]
Going to push out a release for this shortly!

I've managed to get the 'no spaces' validation rule added everywhere customers can be set via email (adding item to cart, updating a cart & checking out).

I didn't however manage to get the customer entry slug's to just be the customer's email address (eg. not slugifying it). It seems that Statamic [slugifies it automatically](https://github.com/statamic/cms/blob/da97be6389b34113a13f3d41ef10b6e95419c2a8/src/Routing/Routable.php#L29) whenever we set the slug on an entry which is a bit of a shame. There's no real way around it, so we're going to have to live with it for now I guess 🤷‍♂️
